tpblite/tpblite.py

```diff
@@ -1,34 +1,64 @@
+from typing import Optional
+
 from .models.torrents import Torrents, Torrent
 from .models.utils import QueryParser
 
-class TPB(object):
-    
-    def __init__(self, base_url='https://tpb.party'):
-        '''ThePirateBay Object
+
+class TPB:
+
+    # PirateBay URL to use for queries
+    base_url: str
```

Comment:
What is the advantage of having this as a class variable type hint? Would it not be more consistent to have as an instance variable?

Reply:
ah, it's actually not a class variable, it's just for type hinting. This syntax (when the class isn't a dataclass) is just used to establish the contract. It means that all TPB objects will have a `base_url` attribute when they are instantiated. You still need to assign it in the init.

This is especially useful as a user of the library, when you use tools like mypy. I'll come back to that in another comment.

Reply:
So, in the absence of a type hinter such as `mypy`, no error will be thrown at runtime if the base_url is not a string?

I'm still not clear on why this is preferred to having this type hint inside of the `__init__` like `__init__(self, base_url: str)`? It seems indicative of being a class variable (even if it is just a type hint and doesn't assign anything per se).

Additionally, since `search_url` is a variable that's been modified by `search` or `browse` methods and not the user directly, is it wise/necessary to have a type hint here? 
Since the `search` or `browse` functions make a request to the url, there's no functionality where the user could, for example, modify `search_url` directly and rerequest. `search_url` really is just an output variable for a sanity check/debugging.

Reply:
> So, in the absence of a type hinter such as mypy, no error will be thrown at runtime if the base_url is not a string?

No error will be thrown at runtime, at least not when creating the TPB object:

```
t = TPB(3)
```

Won't crash immediately (it will of course crash whenever a string operation is triggered on base_url), but it won't crash because the attribute was type-hinted with `base_url: str`.

> I'm still not clear on why this is preferred to having this type hint inside of the __init__ like __init__(self, base_url: str)? It seems indicative of being a class variable (even if it is just a type hint and doesn't assign anything per se).

I just added the type hint to the init function for `base_url`, but this is not related. `base_url` in the init function is a variable you pass to TPB when you instantiate it, but it's not the same as `self.base_url`. Mypy probably won't be able to infer that the var passed to the init becomes the `base_url` attribute. You actually need both type hints: one to warn you if you pass a non-string object to the init, and one to warn you if you try to do a non-string operation on `self.base_url` once you have a TPB object. Please let me know if it's still not clear.

> Additionally, since search_url is a variable that's been modified by search or browse methods and not the user directly, is it wise/necessary to have a type hint here?
> Since the search or browse functions make a request to the url, there's no functionality where the user could, for example, modify search_url directly and rerequest. search_url really is just an output variable for a sanity check/debugging.

I wasn't aware of that. Indeed, if `search_url` is used only in this class and isn't intended to be accessed by the user, it should be a "private" attribute. Since nothing is private in Python, the convention is to prefix it with an underscore: `self._search_url`. I renamed it and moved the type hinting in the init.

Reply:
```python
my_url = "http://google.com"
t = TPB(my_url)  # this is ok

t.base_url += 5  # not ok, Mypy would tell you: can't add int to str

t = TPB(5)  # Not ok, mypy would tell you TPB exepects a str, not an int
```

The snippet above would trigger mypy warnings in your IDE, but you'd still be able to execute the code as if there was no type hinting.

Reply:
```python
    t = TPB(3)

    t = TPB("https://tpb.party")

    t.base_url += 5
```

If you run `mypy tpblite.py`, you would get:

```
tpblite.py:66: error: Argument 1 to "TPB" has incompatible type "int"; expected "str"
tpblite.py:70: error: Unsupported operand types for + ("str" and "int")
```

Reply:
Ahh okay, thanks for the detailed rundown!

> You actually need both type hints: one to warn you if you pass a non-string object to the init, and one to warn you if you try to do a non-string operation on self.base_url once you have a TPB object. Please let me know if it's still not clear.

I've just done a quick test using mypy, and an error is raised correctly for self.base_url, for example
```python
class TPB:

    def __init__(self, base_url: str):
        self.base_url = base_url

    def func(self):
        self.base_url - 3
```
raises the error `Unsupported operand types for - ("str" and "int")`

I think given this, we can omit the extra type hint outside of the `__init__` function

Reply:
You're actually right, mypy can infer the type of some attributes if they are annotated in the init. See the PEP: https://www.python.org/dev/peps/pep-0526/#class-and-instance-variable-annotations.

Annotating an attribute in the body of the class is also valid (see PEP), but seems to be *necessary* only for more complex cases. See https://stackoverflow.com/questions/59478715/type-annotating-class-variable-in-init-or-body/59478877#59478877.

From the SO question:
> The point is that annotating instance variables will always work but inference will sometimes fail.

I think in this case it's simple enough to trust the inference and to not annotate it in the body, even though I'd be in favour of explicit. I removed the type hinting in the body, it's now in init.